### PR TITLE
Feature/baip 1476 xlsx import for options list

### PR DESCRIPTION
**Pull request by bennyvdhoogen**

### CODE REVIEW

File: src/services/import-span-measure-options.ts [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+
+		return formattedData;
+	}
+
+	private fetchMaterialenFromSheet(workSheet, knownSpanMeasureItems?: SpanMeasureItemOption[]) {
+		const minRow = 0;
+		let data: MNummersExcelRowObject[] = xlsx.utils.sheet_to_json<MNummersExcelRowObject>(workSheet, { range: 1 });
+		data = data.slice(minRow <= 2 ? 0 : minRow - 2);
+
+		const formattedData = data.map((row) => {
```
Comment (fortunato) [inline src/services/import-span-measure-options.ts:110]: Useless assignment

File: src/services/import-span-measure-options.ts [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+		this.jsonData = JSON.parse(fs.readFileSync(this.jsonDataFilePath, 'utf8'));
+		const file = await this.getFile();
+
+		const oVSSpanMeasureExcelRowObjectList: OVSSpanMeasureExcelRowObject[] = this.fetchMatrixFromSheet(file.Matrix);
+
+		const fetchBestekposten = this.fetchBestekpostenFromSheet(
+			file.Besteksposten,
+			this.jsonData.spanMeasureItemOptions,
+		);
+		const fetchMaterialen = this.fetchMaterialenFromSheet(file['M-nummers'], this.jsonData.spanMeasureItemOptions);
```
Comment (fortunato) [inline src/services/import-span-measure-options.ts:154]: To me `fetch*` implies an async method that retrieves something using a request.

File: src/services/import-span-measure-options.ts [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+		);
+		this.logger.verbose('Writing JSON to file: ' + filePath);
+
+		fs.writeFileSync(filePath, JSON.stringify(normalizedData), {
+			encoding: 'utf-8',
+		});
+	}
+
+	public formatMaterial(element: string): MeasureItemOption {
+		const material = {
```
Comment (fortunato) [inline src/services/import-span-measure-options.ts:212]: Useless assignment

File: src/services/import-span-measure-options.ts [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+			id: newId(),
+			itemType: SpanMeasureItemType.material,
+			description: element ?? '',
+		};
+
+		return material;
+	}
+
+	public formatSpecificationItem(element: string): MeasureItemOption {
+		const specificationItem = {
```
Comment (fortunato) [inline src/services/import-span-measure-options.ts:222]: Useless assignment

File: src/services/import-span-measure-options.ts [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+
+		return foundItem;
+	}
+
+	public parseSpecificationItems(key: string, list: any): MeasureItemOption[] {
+		if (!key) {
+			return [];
+		}
+
+		const array = key.toString().split(',');
```
Comment (fortunato) [inline src/services/import-span-measure-options.ts:250]: What do you think about chaining this bit too and do with one less assignment?

File: src/services/import-span-measure-options.ts [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+			return [...acc, foundItem];
+		}, []);
+	}
+
+	public parseMaterials(key: string, list: any): MeasureItemOption[] {
+		if (!key) {
+			return [];
+		}
+
+		const itemsSplitByNewline = key.toString().split('\n');
```
Comment (fortunato) [inline src/services/import-span-measure-options.ts:268]: Chain instead of assign?

File: src/services/import-span-measure-options.ts [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+			if (!foundItem) {
+				return acc;
+			}
+
+			return [...acc, foundItem];
+		}, []);
+	}
+
+	private containsMeasureItem(measureOption: MeasureOption, measureItemOption: MeasureItemOption): boolean {
+		return measureOption.measureItems.some((item) => {
```
Comment (fortunato) [inline src/services/import-span-measure-options.ts:282]: Can be written as one-liner without curly braces

File: src/services/import-span-measure-options.ts [hunk 8]
```diff
@@ -0,0 +1,330 @@
...
+	public async normalize(
+		oVSExcelRowObjectList: OVSSpanMeasureExcelRowObject[],
+		materials: object[],
+		bestekposten: object[],
+	): Promise<Record<string, any>> {
+		const measureOptions = {};
+		let lastReadDecompositionType = '';
+
+		oVSExcelRowObjectList.forEach((element: OVSSpanMeasureExcelRowObject, index) => {
+			const foundMeasureOption = this.jsonData.spanMeasureOptions.find((item) => {
```
Comment (fortunato) [inline src/services/import-span-measure-options.ts:296]: Can be written as a one-liner without curly braces

File: src/services/import-span-measure-options.ts [hunk 9]
```diff
@@ -0,0 +1,330 @@
...
+				return item.description == element.Maatregelen;
+			});
+
+			if (element.Maatregelen) {
+				lastReadDecompositionType = element.Onderdelen;
+			}
+
+			const item = {
```
Comment (fortunato) [inline src/services/import-span-measure-options.ts:304]: Arguably a useless assignment

File: src/services/import-span-measure-options.ts [hunk 10]
```diff
@@ -0,0 +1,330 @@
...
+					...this.parseSpecificationItems(element.Besteksposten, bestekposten),
+					...this.parseMaterials(element['Materiaal uit (M)agazijn'], materials),
+				],
+			};
+
+			measureOptions[item.description] = item;
+		});
+
+		// Remap object with named keys to array
+		const measureOptionsArray = Array.from(Object.values(measureOptions));
```
Comment (fortunato) [inline src/services/import-span-measure-options.ts:318]: Useless assignment

File: src/services/import-span-measure-options.ts [hunk 11]
```diff
@@ -0,0 +1,330 @@
...
+
+	private async importOVSMeasures() {
+		this.progressBar.start(100, 0);
+
+		this.jsonData = JSON.parse(fs.readFileSync(this.jsonDataFilePath, 'utf8'));
+		const file = await this.getFile();
+
+		const oVSSpanMeasureExcelRowObjectList: OVSSpanMeasureExcelRowObject[] = this.fetchMatrixFromSheet(file.Matrix);
+
+		const fetchBestekposten = this.fetchBestekpostenFromSheet(
```
Comment (fortunato) [inline src/services/import-span-measure-options.ts:150]: Useless assignments

File: src/services/import-span-measure-options.ts [hunk 12]
```diff
@@ -0,0 +1,330 @@
...
+		return formattedData;
+	}
+
+	private async getFile(): Promise<{ [sheet: string]: WorkSheet }> {
+		const filePath: string =
+			this.configService.get<string>('APP_DIR') +
+			this.configService.get<string>('DOC_DIR') +
+			this.configService.get<string>('READ_FILE_OVS_MEASURES');
+		// read from xlsx file
+		const maxRow = 9628;
```
Comment (fortunato) [inline src/services/import-span-measure-options.ts:127]: Is it worth adding a comment on how this number is relevant?

File: src/services/import-span-measure-options.ts [hunk 13]
```diff
@@ -0,0 +1,330 @@
+import * as fs from 'fs';
+import * as path from 'path';
+
```
Comment (fortunato) [inline src/services/import-span-measure-options.ts:3]: Why the newline here?

File: src/services/import-span-measure-options.spec.ts [hunk 14]
```diff
@@ -0,0 +1,31 @@
+import fs from 'fs';
+
```
Comment (fortunato) [inline src/services/import-span-measure-options.spec.ts:2]: Why the newline here?

File: src/services/import-span-measure-options.spec.ts [hunk 15]
```diff
@@ -0,0 +1,31 @@
+import fs from 'fs';
```
Comment (fortunato) [inline src/services/import-span-measure-options.spec.ts:1]: Nice one :ok_hand:

**Review by fortunato — CHANGES_REQUESTED** [review]
Looks neat :+1: 

A couple of minor remarks.

File: src/services/import-span-measure-options.ts (hunk 2 above)
Comment (bennyvdhoogen) [reply src/services/import-span-measure-options.ts]: Agree with that one, renamed these to 'get...'

File: src/services/import-span-measure-options.ts [hunk 16]
```diff
@@ -0,0 +1,330 @@
+import * as fs from 'fs';
```
[the remark this answers is not in the document]
Comment (bennyvdhoogen) [reply src/services/import-span-measure-options.ts:1]: I would be in favor if that is possible. Most frameworks I have worked with use a folder/namespace for CLI command such as this, but I think having stuff like this in a specific module aligns with the notion that we split our code based on a different domain. Will try to move this stuff to the `span-installation` folder -- although we had our thoughts about that one I guess? Perhaps better to do make a small refactor ticket for ths?

**Review by fortunato — APPROVED** [review]

**Pull request merged**